### PR TITLE
Support unpacked array port in protect-lib and hierarchical verilation

### LINKED ISSUE #2558: The --hierarchical option causes failure with "--make gmake" option: Unsupported: unpacked arrays with protect-lib on 'val'
```
%Error-UNSUPPORTED: x.v:36:16: Unsupported: unpacked arrays with protect-lib on 'val'
   36 | input[31:0]  val[5:0];
      |              ^~~~~~
```

This failure doesn't happen without the ```--hierarchical``` option, and also with the ```--hierarchical``` option with ```--make cmake```.

### PR DESCRIPTION
**Pull request by yTakatsukasa**
#2670 and #2671 are needed by this PR.
Closes #2558 (Though the issue is talking about gmake, unpacked array hasn't been supported in protect-lib and hierarchical verilation regardless of build tool).

Because this is not a cleanup PR, please review this.

### CODE REVIEW

File: src/V3EmitV.cpp [hunk 1]
```diff
@@ -629,11 +629,30 @@ class EmitVBaseVisitor VL_NOT_FINAL : public EmitCBaseVisitor {
...
-        putfs(nodep, nodep->verilogKwd());
-        puts(" ");
-        iterate(nodep->dtypep());
-        puts(" ");
-        puts(nodep->prettyName());
+        if (nodep->isIO()) {
+            putfs(nodep, nodep->verilogKwd());
+            puts(" ");
+        }
+        std::vector<AstUnpackArrayDType*> unpackps;
```
Comment (wsnyder) [inline src/V3EmitV.cpp:636]:
```suggestion
        std::vector<const AstUnpackArrayDType*> unpackps;
```

File: src/V3EmitV.cpp [hunk 2]
```diff
@@ -629,11 +629,30 @@ class EmitVBaseVisitor VL_NOT_FINAL : public EmitCBaseVisitor {
...
+        for (AstNodeDType* dtypep = nodep->dtypep(); dtypep;) {
+            if (AstUnpackArrayDType* unpackp = VN_CAST(dtypep, UnpackArrayDType)) {
```
Comment (wsnyder) [inline src/V3EmitV.cpp:638]:
```suggestion
        for (AstNodeDType* dtypep = nodep->dtypep(); dtypep;) {
            dtypep = dtypep->skipRefp();
            if (AstUnpackArrayDType* unpackp = VN_CAST(dtypep, UnpackArrayDType)) {
```

And have your test use.reference through a typedef.

File: src/V3EmitV.cpp [hunk 3]
```diff
@@ -629,11 +629,30 @@ class EmitVBaseVisitor VL_NOT_FINAL : public EmitCBaseVisitor {
...
+                dtypep = unpackp->subDTypep();
+            } else {
+                iterate(dtypep);
+                puts(" ");
+                puts(nodep->prettyName());
+                dtypep = nullptr;
+            }
+        }
+        // If nodep is an unpacked array, append unpacked dimensions
+        for (auto unpackp : unpackps) {
```
Comment (wsnyder) [inline src/V3EmitV.cpp:649]:
```suggestion
        for (const auto& unpackp : unpackps) {
```

**Review by wsnyder — COMMENTED** [review]
Does this work with 2-D unpacked?  I suspect not, but perhaps easy to get working?  If not may need to resurrect t_prot_lib_unpacked_bad.v

**Comment by wsnyder** [comment]
Feel free to merge when happy

**Comment by yTakatsukasa** [comment]
Thanks for the review.
2D array is working. I'd like to add some more tests, then merge later.

**Comment by yTakatsukasa** [comment]
Some 2D and 3D array test is added.
I will merge tomorrow if CI passes.

**Comment by yTakatsukasa** [comment]
Hmm,  a02bafd73dcc1d4453b3c3736b1f7c596bfc3ed2 passes.
The change after the commit is just expanding t_prot_lib test, but t_debug_emitv fails in c2b5f15adb92f3c4383ca6111e905f9f49d038a3.

I confirmed the test fails after merging master. Maybe something merged to master recently caused this failure.
I will check further.

**Comment by wsnyder** [comment]
I did recently change debug_emitv so maybe it's astable.

**Comment by yTakatsukasa** [comment]
Thanks.
I simply update the golden. Hopefully the test will pass.

**Pull request merged**